### DISCUSSION TITLE
[Deployment] Deploying Token contract with Create2Factory

### DISCUSSION THREAD

**Discussion by anna-sve** [zkSync-Community-Hub/zksync-developers#717] [Deployments] [answered]
### Environment

Mainnet

### zkSolc Version

latest: v1.5.4

### zksync-ethers Version

6.13.0

### Minimal Reproducible Code or Repo Link

```
const { Wallet, Provider, ContractFactory } = require("zksync-ethers");
const { ethers } = require("hardhat");
require("dotenv").config();

const provider = Provider.getDefaultProvider();
const abiJson = require("../artifacts-zk/contracts/erc20/At.sol/ATokenA.json");
const byteCode = abiJson.bytecode;
const abi = abiJson.abi;
const wallet = new Wallet(process.env.WALLET_PRIVATE_KEY as string, provider);
const factory = new ContractFactory(abi, byteCode, wallet);

const name = "ATokenA";
const symbol = "AT";
const totalSupply = ethers.parseUnits("1000000000000", 18); // 1 trillion tokens
const decimals = 18;
const dust = ethers.parseUnits("1", 16); // 0.01 tokens as dust

export default async function deployContractWithArgs(name: string, symbol: string, totalSupply: number, decimals: number, dust: number) {
 
const salt = ethers.id("AT-Salt-515");

  const contract = await factory.getDeployTransaction(name, symbol, totalSupply, decimals, dust, {salt});
  await contract.deployed();
  console.log("Contract deployed to:", contract.address);
}

deployContractWithArgs(name, symbol, totalSupply, decimals, dust)
  .then(() => process.exit(0))
  .catch((error) => {
    console.error(error);
    process.exit(1);
  });
  ```
  

### Does this work on other EVMs? (If yes, please list at least 1 of them)

ETH

### Description of What Your Contract Does

Trying to ERC20 token contract deployment with the instructions given;
https://sdk.zksync.io/js/ethers/api/v6/contract/contract-factory

### Additional Details

TypeError: contract.deployed is not a function
    at /Users/anna/code/projects/atoken/eth-zksync-deploy/deploy/deterministicDeploy2.ts:23:18
    at step (/Users/anna/code/projects/atoken/eth-zksync-deploy/deploy/deterministicDeploy2.ts:33:23)
    at Object.next (/Users/anna/code/projects/atoken/eth-zksync-deploy/deploy/deterministicDeploy2.ts:14:53)
    at fulfilled (/Users/anna/code/projects/atoken/eth-zksync-deploy/deploy/deterministicDeploy2.ts:5:58)
    
    ### also not success with contract.waitForDeployment()

**Comment by alhenaai** [comment, marked as answer]
It seems like you're encountering an issue with the `contract.deployed()` function not being recognized in your deployment script for an ERC20 token on ZKsync using `zksync-ethers`. This issue might be due to the differences in how contract deployment is handled in ZKsync compared to Ethereum.

In ZKsync, the deployment process does not use the `deployed()` method as typically seen in Ethereum's Ethers.js. Instead, you should be using the `getDeployTransaction()` method from `ContractFactory`, and then sending this transaction to the network. Here’s how you can adjust your deployment script:

1. **Generate the Deployment Transaction**: You've correctly used `getDeployTransaction()` to create the deployment transaction. This is the right approach for ZKsync.
2. **Send the Transaction**: Instead of calling `deployed()`, you should send the transaction to the network directly using the wallet. For example:
   ```javascript
   const deployTx = factory.getDeployTransaction(name, symbol, totalSupply, decimals, dust, {salt});
   const response = await wallet.sendTransaction(deployTx);
   await response.wait(); // Wait for the transaction to be mined
   console.log("Contract deployed to:", response.contractAddress);
   ```

3. **Check Deployment Status**: After sending the transaction, you can check the status by waiting for the transaction to be mined using `response.wait()` and then logging the contract address from the response.

Here’s a revised version of your deployment function:
```javascript
export default async function deployContractWithArgs(name, symbol, totalSupply, decimals, dust) {
  const salt = ethers.id("AT-Salt-515");
  const deployTx = factory.getDeployTransaction(name, symbol, totalSupply, decimals, dust, {salt});
  const response = await wallet.sendTransaction(deployTx);
  await response.wait(); // This replaces the deployed() method
  console.log("Contract deployed to:", response.contractAddress);
}
```

This adjustment should align your deployment script with the capabilities of the `zksync-ethers` library and the ZKsync environment. Make sure your environment variables and network configurations are correctly set to connect to the ZKsync mainnet. If you continue to face issues, reviewing the [ZKsync SDK documentation](https://sdk.zksync.io/js/ethers/api/v6/contract/contract-factory) or the specific error messages can provide further guidance.

Your feedback helps us serve you better. If this fully addresses your question, please give it a thumbs up. If it doesn't, please give it a thumbs down. This will inform the community and team so we can improve our support.
